src/main/java/org/opensearch/knn/index/codec/native_index/NativeIndexBuilder.java

```diff
@@ -0,0 +1,177 @@
+        throws IOException {
+        NativeIndexInfo indexInfo = getIndexInfo(fieldInfo, valuesProducer, indexPath);
+        BinaryDocValues values = valuesProducer.getBinary(fieldInfo);
+        if (isMerge) {
+            startMergeStats(indexInfo.numDocs, indexInfo.arraySize);
+        }
+        if (isRefresh) {
+            recordRefreshStats();
+        }
+        createIndex(indexInfo, values);
+        if (isMerge) {
+            endMergeStats(indexInfo.numDocs, indexInfo.arraySize);
+        }
+    }
+
+    /**
+     * Method that makes a native index given the parameters from indexInfo
+     * @param indexInfo
+     * @param values
+     * @throws IOException
+     */
+    protected abstract void createIndex(NativeIndexInfo indexInfo, BinaryDocValues values) throws IOException;
+
+    /**
+     * Method that generates extra index parameters to be passed to the native library
+     * @param fieldInfo
+     * @param knnEngine
+     * @return
+     * @throws IOException
+     */
+    protected abstract Map<String, Object> getParameters(FieldInfo fieldInfo, KNNEngine knnEngine) throws IOException;
+
+    /**
+     * Method that gets the native vector info
+     * @param fieldInfo
+     * @param testValues
+     * @return
+     * @throws IOException
+     */
+    protected abstract NativeVectorInfo getVectorInfo(FieldInfo fieldInfo, BinaryDocValues testValues) throws IOException;
+
+    protected VectorTransfer getVectorTransfer(VectorDataType vectorDataType) {
+        if (VectorDataType.BINARY == vectorDataType) {
+            return new VectorTransferByte(KNNSettings.getVectorStreamingMemoryLimit().getBytes());
+        }
+        return new VectorTransferFloat(KNNSettings.getVectorStreamingMemoryLimit().getBytes());
+    }
+
+    /**
+     * Method that gets the native index info from a given field
+     * @param fieldInfo
+     * @param valuesProducer
+     * @param indexPath
+     * @return
+     * @throws IOException
+     */
+    private NativeIndexInfo getIndexInfo(FieldInfo fieldInfo, DocValuesProducer valuesProducer, String indexPath) throws IOException {
+        BinaryDocValues testValues = valuesProducer.getBinary(fieldInfo);
+        NativeIndexInfo indexInfo = new NativeIndexInfo();
+        indexInfo.fieldInfo = fieldInfo;
+        indexInfo.knnEngine = getKNNEngine(fieldInfo);
+        indexInfo.numDocs = (int) KNNCodecUtil.getTotalLiveDocsCount(testValues);
+        indexInfo.vectorInfo = getVectorInfo(fieldInfo, testValues);
+        indexInfo.arraySize = indexInfo.numDocs * getVectorSize(indexInfo.vectorInfo);
+        indexInfo.parameters = getParameters(fieldInfo, indexInfo.knnEngine);
+        indexInfo.indexPath = indexPath;
+        return indexInfo;
+    }
+
+    private long getVectorSize(NativeVectorInfo vectorInfo) {
+        if (vectorInfo.vectorDataType == VectorDataType.BINARY) {
+            return vectorInfo.dimension / 8;
+        } else {
+            return vectorInfo.dimension * 4;
+        }
+    }
```

Comment:
interesting I think you want the bytes for a single vector here. I would rename this to getBytesPerVector